doc/man/wesnothd.6
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"
.
.TH WESNOTHD 6 "2022" "wesnothd" "Battle for Wesnoth multiplayer network daemon"
.
.SH NAME
.
wesnothd \- Battle for
.B Wesnoth
multiplayer network daemon
.
.SH SYNOPSIS
.
.B wesnothd
.RB [\| \-dvw \|]
.RB [\| \-c
.IR path \|]
.RB [\| \-p
.IR port \|]
.br
.B wesnothd
.B \-V
.
.SH DESCRIPTION
.
Manages Battle for Wesnoth multiplayer games. See https://www.wesnoth.org/wiki/ServerAdministration
on what commands the server accepts via the wesnoth client (/query ...) or the fifo.
.
.SH OPTIONS
.
.TP
.BI -c\  path ,\ --config \ path
tells wesnothd where to find the config file to use. See the section
.B SERVER CONFIG
below for the syntax. You can reload the config with sending SIGHUP to the server process.
.TP
.B -d, --daemon
runs wesnothd as a daemon.
.TP
.B -h, --help
tells you what the command line options do.
.TP
.BI --log- level = domain1 , domain2 , ...
sets the severity level of the log domains.
.B all
can be used to match
any log domain. Available levels:
.BR error ,\  warning ,\  info ,\  debug .
By default the
.B error
level is used and the
.B info
level for the
.B server
domain.
.TP
.BI -p\  port ,\ --port \ port
binds the server to the specified port. If no port is specified, port
.B 15000
will be used.
.TP
.B -V, --version
shows version number and exits.
.TP
.B -v, --verbose
turns debug logging on.
.
.SH SERVER CONFIG
.
.SS The general syntax is:
.nf
\&
\fIglobal_key\fR="\fIvalue\fR"
[\fItag\fR]
.in +4n
\fIkey\fR="\fIvalue\fR"
.in -4n
[/\fItag\fR]
.in -4n
.fi
.SS "Global keys:"
.
.TP
.B allow_remote_shutdown
If set to \fBno\fR (default), shut_down and restart requests are ignored unless they come from the fifo.
Set it to \fByes\fR to allow remote shutdown via a /query by an administrator.
.TP
.B ban_save_file
Full or relative path to a (gzip compressed) file that the server can read and write.
Bans will be saved to this file and read again on server start.
.TP
.B connections_allowed
The number of allowed connections from the same IP. \fB0\fR means infinite. (default: \fB5\fR)
.TP
.B deny_unregistered_login
Whether to prevent users who are not registered with the
.B user_handler
being used from logging in. (default: false)
.TP
.B disallow_names
Names/nicks that are not accepted by the server. \fB*\fR and \fB?\fR from wildcard patterns are supported. See
.BR glob (7)
for more details.
Default values (used if nothing is specified) are:
.BR *admin*,*admln*,*moderator*,*server*,ai,ai?,computer,human,network,player .
.TP
.B fifo_path
The path to the fifo you can echo server commands into (same as /query ... from wesnoth).
If not specified defaults to the compile-time path (default: \fB/var/run/wesnothd/socket\fR).
.TP
.B id
The ID of the server when using the database functionality to store game information. Generally is the major.minor version of accepted clients, such as 1.16 or 1.17.
.TP
.B max_messages
The number of allowed messages in \fBmessages_time_period\fR. (default: \fB4\fR)
.TP
.B messages_time_period
The time period (in seconds) message flooding is detected in. (default: \fB10\fR seconds)
.TP
.B motd
The message of the day.
.TP
.B passwd
The password used to gain admin privileges (via \fB/query admin \fIpassword\fR).
.TP
.B replay_save_path
The directory where the server stores game replays. (Don't forget the trailing /!) Defaults to `' which means the directory wesnothd was started in.
.TP
.B restart_command
The command that the server uses to start a new server process via the \fBrestart\fR command. (Can only be issued via the fifo. See the \fBallow_remote_shutdown\fR setting.)
.TP
.B save_replays
Defines whether the server will automatically save replays of games. (default: \fBfalse\fR)
.TP
.B tls_enabled
Enable use of TLS. If true, requires 3 following options to be set too.
.TP
.B tls_fullchain
TLS full certificate chain in .pem format
.TP
.B tls_private_key
private key for TLS in .pem format
.TP
.B tls_dh
TLS DH parameters, as .pem file. Generate with \fBopenssl dhparam -out filename.pem 4096\fR
.TP
.B versions_accepted
A comma separated list of version strings to be accepted by the server. \fB*\fR and \fB?\fR from wildcard patterns are supported.
(defaults to the corresponding wesnoth version)
.br
Example: \fBversions_accepted="*"\fR accepts any version string.
.
.SS "Global tags:"
.
.P
.B [redirect]
A tag to specify a server to redirect certain client versions to. Is not used if
.B versions_accepted
is not set.
.RS
.TP
.B host
The address of the server to redirect to.
.TP
.B port
The port to connect to.
.TP
.B version
A comma separated list of versions to redirect. Behaves the same way as
.B versions_accepted
in regard to wildcard patterns.
.RE
.P
.B [ban_time]
A tag to define convenient keywords for temporary ban time lengths.
.RS
.TP
.B name
The name used to reference the ban time.
.TP
.B time
The time length definition.
The format is: %d[%s[%d%s[...]]] where %s is s (seconds), m (minutes), h (hours), D (days), M (months) or Y (years) and %d is a number.
If no time modifier is given minutes (m) are assumed.
Example:
.B time="1D12h30m"
results in a ban time of 1 day, 12 hours and 30 minutes.
.RE
.P
.B [proxy]
A tag to tell the server to act as a proxy and forward the connected client's requests to the specified server.
Accepts the same keys as
.BR [redirect] .
.RE
.P
.B [user_handler]
Configures the user handler. If no
.B [user_handler]
section is present in the configuration the server will run without any nick registration service. All additional tables that are needed for the
.B forum_user_handler
to function can be found in table_definitions.sql in the Wesnoth source repository. Requires mysql support enabled. For cmake this is
.B ENABLE_MYSQL
and for scons this is
.B forum_user_handler.
.RS
.TP
.B db_host
The hostname of the database server
.TP
.B db_name
The name of the database
.TP
.B db_user
The name of the user under which to log into the database
.TP
.B db_password
This user's password
.TP
.B db_users_table
The name of the table in which your phpbb forum saves its user data. Most likely this will be <table-prefix>_users (e.g. phpbb3_users).
.TP
.B db_extra_table
The name of the table in which wesnothd will save its own data about users.
.TP
.B db_game_info_table
The name of the table in which wesnothd will save its own data about games.
.TP
.B db_game_player_info_table
The name of the table in which wesnothd will save its own data about the players in a game.
.TP
.B db_game_content_info_table
The name of the table in which wesnothd will save its own data about game content (scenario/era/modifications) used in a game.
.TP
.B db_user_group_table
The name of the table in which your phpbb forum saves its user group data. Most likely this will be <table-prefix>_user_group  (e.g. phpbb3_user_group).
.TP
.B db_tournament_query
The SQL query to find tournaments to announce on login. Should return tournament \fBtitle\fR, \fBstatus\fR and \fBurl\fR.
.TP
.B db_connection_history_table
The name of the table in which to store login/logout times. Also used for matching IPs to users and vice versa.
.TP
.B db_topics_table
The name of the table in which your phpbb forum saves its topic (thread) information. Most likely this will be <table-prefix>_topics  (e.g. phpbb3_topics).
.TP
.B db_banlist_table
The name of the table in which your phpbb forum saves its user bans data. Most likely this will be <table-prefix>_banlist  (e.g. phpbb3_banlist).
.TP
.B mp_mod_group
The ID of the forum group to be considered as having moderation authority.
.TP
.B forum_admin_group
The ID of one of the forum groups to be considered as having admin authority on the add-ons server.
.TP
.B site_admin_group
The ID of one of the forum groups to be considered as having admin authority on the add-ons server.
.RE
.
.SH EXIT STATUS
.
Normal exit status is 0 when the server was properly shutdown. An exit status of 2 indicates an error with the command line options.
.
.SH AUTHOR
.
Written by David White <davidnwhite@verizon.net>.
Edited by Nils Kneuper <crazy-ivanovic@gmx.net>, ott <ott@gaon.net>, Soliton <soliton.de@gmail.com> and Thomas Baumhauer <thomas.baumhauer@gmail.com>.
This manual page was originally written by Cyril Bouthors <cyril@bouthors.org>.
.br
Visit the official homepage: https://www.wesnoth.org/
.
.SH COPYRIGHT
.
Copyright \(co 2003-2025 David White <davidnwhite@verizon.net>
.br
This is Free Software; this software is licensed under the GPL version 2, as published by the Free Software Foundation.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.
.SH SEE ALSO
.
.BR wesnoth (6)



changelog
Version 1.13.11+dev:
 * Add-ons client:
   * Addon upload progress bar now works also on Windows (bug #1439).
   * Fixed inability to cancel addon upload (bug #2591).
 * Add-ons server:
   * Added support for adding free-form comma-separated tags to add-ons in
     their publishing info (bug #2565).
   * Added support for overriding existing add-on attributes using the control
     FIFO.
 * Campaigns:
   * Northern Rebirth:
     * Fixed S06a Rakshas displaying incorrect portrait (issue #2569)
     * Fixed S12a invalid side error (issue #2569)
     * Fixed S13a incorrect leader when Eryssa is alive (issue #2569)
     * Fix Krash being brought back to life if he is dead
     * Appearance improvements for S04a 06a, 07a, 09a, 10a, 11a, 12a, 13a
     * Fix gold storing in S07a for S13a if retrieved
     * S04a, 05a, & S13a Make enemy units loyal so gold & events work properly
 * Language and i18n:
   * Updated translations: British English, Spanish
 * Lua API:
   * Add side_name to unit proxy
 * User Interface:
   * Improved outro screen.
   * Fixed a few cases of data not displaying in the MP Join Game screen.
   * Fixed required addon names not displaying properly when joining an
     MP game.
   * The Toggle Fullscreen hotkey now works everywhere.
 * WML engine:
   * Fixed units shown with [move_units_fake] disappearing between steps
     (bug #1516).
   * [modify_side] now supports side_name
   * [set_menu_item] no longer fires repeatedly if the player holds the
     hotkey (bug #1711). If you were relying on repeated firing, add
     repeat_on_hold=yes to [default_hotkey].
   * [set_variable] now supports prefix and suffix operations for
     string concatenation.
   * [effect]apply_to=attack now supports set_range
 * Miscellaneous and bug fixes:
   * Fixed standing animation toggle not taking immediate effect (bug
     #1653).
   * Fixed error when de-assigning village with [capture_village].
   * Enabled the use of Open Type Font (.otf) fonts.
   * Fresh installs of the game will now open in fullscreen by default.
   * New game theme music by Mattias Westlund.
   * The unit advancement prompt is no longer shown for droided sides.
   * Fixed custom game titles being lost when reloading MP games.
   * The copy-to-clipboard function on the Lua console now produces plain text
     without Pango markup or entities for special characters (<, > and &)
     (bug #2434).
   * 'Turn Changed' desktop notifications in MP will now only display when
     human-controlled sides take control.
 * Music and sound effects:
   * Updated a few UI sounds.

Version 1.13.11:
 * Add-ons client:
   * Added an order dropdown that allows you to sort add-ons by the time of
     latest update or original upload (issue #1747)
   * Players will now be prompted to update outdated dependencies alongside
     downloading missing ones when installing an add-on.
 * Add-ons server:
   * Empty passphrases from malfunctioning clients that do not provide or
     generate a passphrase otherwise are now rejected instead of treated as
     valid.
 * Campaigns:
   * The Cutscene_Minimal theme is now used in all dialog-only scenarios that
     have `linger=no` in [end_level].
   * An Orcish Incursion:
     * New story art.
   * A Tale of Two Brothers:
     * Added a full set of hints on easy difficulty.
   * Delfador's Memoirs:
     * Iliah-Malal can now be killed by either undead or Delfador in S19
     * S19 has been rebalanced to increase difficulty and allow player & enemy
       to field more units.
   * Descent into Darkness:
     * 'Endless Night' now has accurate & improved scenario objectives. It also
       has received improvements in dialog, map appearance, and difficulty.
     * S08/S09 A Small Favor 2/3 - Fixed bug preventing door operation/usage.
     * Fixed and improved appearance and handling of the wose.
     * Other minor fixes and improvements.
 * Packaging
   * Renamed the target and binary for the Boost unit tests from `test` to
     `boost_unit_tests`. This quells warnings from CMake about reserved target
     names, and reduces confusion about which `test` is intended: the wesnoth
     executable or the standard shell command.
   * The CMake build system now uses standard Kitware-supported variables
     from the `GNUInstallDirs` module. Distributions might have to change
     the defaults to suit their FHS needs.
   * The previous addition of `GNUInstallDirs` necessitates a bumping of the
     CMake minimum requirements, which is now at 2.8.5. Most distributions
     already use a CMake version above 3.0, so this would only affect users
     of vintage CMake versions.
   * higher resolutions of the application icon have been added and are now
     stored in the hicolor icon theme's directory
 * Language and i18n:
   * Updated translations: British English, Chinese (Simplified), Czech,
     Italian, Spanish
 * Lua API:
   * New wesnoth.unit_types[] getters:
     * `advances_to`
     * `advances_from`
     * `profile`
     * `small_profile`
   * The wml module now offers another new way of accessing WML variables:
     wml.variables is a simple wrapper around wml.get|set_variable.
   * unit.id is now a modifiable field for off-map (Lua-only) units.
   * Allow moving on-map units by setting `unit.x` and `unit.y` fields, or with
     `unit.loc = {x, y}` or `unit.loc = {x = x, y = y}`
   * Added `side.chose_random` getter.
   * Lua GUI2 API: added support for slider callbacks via
     `wesnoth.set_dialog_callback`.
 * Multiplayer:
   * Dark Forecast: Fixed broken faction and leader selection.
   * Rename the Khalifate to Dunefolk. This includes renaming all the faction's
     units.
 * Performance:
   * GUI2 windows no longer redraw everything 50 times per second. This reduces
     CPU usage in fullscreen windows such as MP lobby by about 85 %.
   * Miscellaneous low-level optimizations in game rendering code, improving
     performance ingame by up to 50 %.
 * Units:

utils/wesnothd.pl
use strict;
use wml;
use IO::Socket;
use IO::Select;
use Fcntl;
use Carp qw(confess);
use Getopt::Std;

#my $usage = "Usage: $0 [-p port]";

my %opts = ();
getopts('p:',\%opts);

my $port = '15000';
$port    = $opts{'p'} if $opts{'p'};

my %pending_handshake = ();
my %outgoing_schemas = ();
my %incoming_schemas = ();
my %incoming_bufs = ();
my %connection_nums = ();
my %user_name_socket = ();
my %socket_user_name = ();
my %lobby_players = ();
my %socks_map = ();
my @games = ();
my $current_game_id = 1;
my $current_connection_num = 1;
my $read_set = new IO::Select();

my $login_response = &wml::single_child_doc('mustlogin');
my $join_lobby_response = &wml::single_child_doc('join_lobby');
my $gamelist =  {'name' => 'gamelist', 'children' => [], 'attr' => {}};
my $initial_response = {'name' => '', 'children' => [$gamelist], 'attr' => {}};
my $old_initial_response = &wml::deep_copy($initial_response);

my @disallowed_names = ('server', 'ai', 'player', 'network', 'human', 'admin', 'computer');

my $server_sock = new IO::Socket::INET (
										LocalHost => '',
										LocalPort => $port,
										Proto => 'tcp',
										Listen => 1,
										Reuse => 1
									   ) or die "could not create server sock: $!";

$read_set->add($server_sock);

sub send_data {
	my $doc = shift @_;
	return unless @_;
	print "SENDING: {{{" . &wml::write_text($doc) . "}}}\n";
	foreach my $socket (@_) {
		my $sock = $socks_map{$socket};
		my $packet = &wml::write_binary($outgoing_schemas{$sock}, $doc) . chr 0;
		my $header = pack('N',length $packet);
		$packet = $header . $packet;
		my $res = syswrite($sock, $packet, length $packet);
		&socket_disconnected($sock) unless $res == length $packet;
	}
}

sub send_error($$) {
	my ($sock, $msg) = @_;
	print STDERR "sending error: '$msg'\n";
	my $doc = {'name' => '', 'children' => [{'name' => 'error', 'attr' => {'message' => $msg}}], 'attr' => {}};
	&send_data($doc,$sock);
}

sub send_to_lobby {
	my ($doc, $exclude, @users) = @_;
	foreach my $socket (@users) {
		&send_data($doc, $socks_map{$socket}) unless $socks_map{$socket} eq $exclude;
	}
}

sub socket_connected($) {
	my ($sock) = @_;
	print STDERR "connected: $sock\n";
	$read_set->add($sock);
	$outgoing_schemas{$sock} = [];
	$incoming_schemas{$sock} = [];
	$pending_handshake{$sock} = 1;
	$socks_map{$sock} = $sock;
	my $flags = '';
	fcntl($sock, F_GETFL, $flags) or die "Couldn't get flags: $!";
	$flags |= O_NONBLOCK;
	fcntl($sock, F_SETFL, $flags) or die "Couldn't set flags: $!";
}

sub socket_disconnected($) {
	my ($sock) = @_;
	$sock = $socks_map{$sock};
	print STDERR "disconnected: $sock\n";
	delete $socks_map{$sock};
	delete $outgoing_schemas{$sock};
	delete $incoming_schemas{$sock};
	delete $incoming_bufs{$sock};
	delete $pending_handshake{$sock};
	delete $connection_nums{$sock};
	delete $lobby_players{$sock};
	if(my $username = $socket_user_name{$sock}) {
		delete $user_name_socket{$username};
		delete $socket_user_name{$sock};
		print STDERR "searching for '$username'...\n";
		my $users = $initial_response->{'children'};
		for(my $n = 0; $n != @$users; ++$n) {
			my $user = $users->[$n];
			my $attr = $user->{'attr'};
			if($user->{'name'} eq 'user' and $attr->{'name'} eq $username) {
				my $len = @$users;
				print STDERR "OLD USERS: " . (join ',', @$users) . "\n";
				print STDERR "removing item $n...\n";
				my @new_users = ();
				for(my $m = 0; $m != @$users; ++$m) {
					push @new_users, $users->[$m] unless $m == $n;
				}
				@$users = @new_users;
				print STDERR "NEW USERS: " . (join ',', @$users) . "\n";
				last;
			}
		}

		&sync_lobby();
	}
	$read_set->remove($sock);
	close $sock;
}

sub sync_lobby {
	my $diff = &wml::get_diff($initial_response, $old_initial_response, 'gamelist_diff');
	while(my $sock = each %lobby_players) {
		&send_data($diff, $sock);
	}

	$old_initial_response = &wml::deep_copy($initial_response);
}

sub received_packet($$) {
	my ($sock, $packet) = @_;
	printf STDERR "received %d from $sock\n", length $packet;
	chop $packet;
	my $doc = &wml::read_binary($incoming_schemas{$sock}, $packet);
	my $text = &wml::write_text($doc);
	print STDERR "RECEIVED: $text";

	if(not $socket_user_name{$sock}) {
		if(my $login = &wml::has_child($doc, 'login')) {
			my $attr = $login->{'attr'};
			my $username = $attr->{'username'};
			print STDERR "user attempting to log in as \"$username\"...\n";
			unless($username =~ /^[A-Za-z0-9_]{1,18}$/) {
				&send_error($sock, 'This username is not valid.');
				return;
			}

			foreach my $disallowed (@disallowed_names) {
				if($disallowed eq $username) {
					&send_error($sock, 'This username is disallowed.');
					return;
				}
			}

			if($user_name_socket{$username}) {
				&send_error($sock, 'This username is already taken.');
				return;
			}

			print STDERR "log in okay; telling to join lobby\n";
			&send_data($join_lobby_response, $sock);

			my $users = $initial_response->{'children'};
			push @$users, {'name' => 'user', 'children' => [], 'attr' => {'name' => $username, 'available' => 'yes'}};
			&send_data($initial_response, $sock);
			&sync_lobby();
			$lobby_players{$sock} = 1;
			$user_name_socket{$username} = $sock;
			$socket_user_name{$sock} = $username;
			return;
		} else {
			&send_error($sock, 'You must login first.');
			return;
		}
	} elsif($lobby_players{$sock}) {
		if(my $create_game = &wml::has_child($doc, 'create_game')) {
			$create_game->{'name'} = 'game';
			push @games, &create_game($create_game, $sock);
			delete $lobby_players{$sock};
		} elsif(my $message = &wml::has_child($doc, 'message')) {
			my $attr = $message->{'attr'};
			$attr->{'sender'} = $socket_user_name{$sock};
			&send_to_lobby($doc, $sock, keys %lobby_players);
		} elsif(my $whisper = &wml::has_child($doc, 'whisper')) {
			my $attr = $whisper->{'attr'};
			$attr->{'sender'} = $socket_user_name{$sock};
			my $receiver = $attr->{'receiver'};
			&send_data($doc, $user_name_socket{$receiver});
		#} elsif(my $query = &wml::has_child($doc, 'query')) {
		#	my $type = $query->{'attr'}->{'type'};
		}
		return;
	} else {
		my $game;
		foreach my $g (@games) {
			my $players = $g->{'players'};
			if($players->{$sock}) {
				$game = $g;
				last;
			}
		}

		print STDERR "Illegal socket: '$sock'\n" and return unless $game;
		if($game->{'owner'} eq $sock and
		   my $side = &wml::has_child($doc, 'side')) {
			if(not &wml::has_child($game->{'level'},'side')) {
				#level is not initialized yet, initialize it
				my $level = $game->{'level'};
				my $level_attr = $level->{'attr'};
				my $doc_attr = $doc->{'attr'};
				if($side->{'shroud'} ne 'yes') {
					$level_attr->{'map_data'} = $doc_attr->{'map_data'};
					$level_attr->{'map'} = $doc_attr->{'map'};
				}
				if(my $era = &wml::has_child($doc, 'era')) {
					$level_attr->{'mp_era'} = $era->{'attr'}->{'id'};
				}

				my @mp_flags = qw(mp_use_map_settings mp_village_gold mp_fog mp_shroud experience_modifier mp_countdown mp_countdown_init_time mp_countdown_turn_bonus mp_countdown_reservoir_time mp_countdown_action_bonus hash);
				foreach my $flag (@mp_flags) {
					$level_attr->{$flag} = $doc_attr->{$flag};
				}

				my $games = $gamelist->{'children'};
				push @$games, $level;
				$game->{'description'} = $level;
				$game->{'level'} = $doc;

				&sync_lobby();
			} else {
			}
		}
	}
}

sub handshake($) {
	my ($sock) = @_;
	my $buf = '';
	my $res = sysread($sock, $buf, 4);
	if($res != 4) {
		&socket_disconnected($sock);
	} else {
		my $buf = pack('N', $current_connection_num);
		$connection_nums{$sock} = $current_connection_num++;
		my $res = syswrite($sock, $buf, 4);
		if($res != 4) {
			&socket_disconnected($sock);
		} else {
			delete $pending_handshake{$sock};
			&send_data($login_response, $sock);
		}
	}
}

sub create_game
{
	my ($level, $owner) = @_;
	my $attr = $level->{'attr'};
	$attr->{'id'} = $current_game_id;
	{'owner' => $owner, 'members' => [$owner], 'players' => {$owner => 1}, 'level' => $level, 'id' => $current_game_id++};
}

while(1) {
	my ($rh_set) = IO::Select->select($read_set, undef, undef, 1000);
	foreach my $sock (@$rh_set) {
		if($sock == $server_sock) {
			my $new_sock = $sock->accept();
			&socket_connected($new_sock);
		} elsif($pending_handshake{$sock}) {
			&handshake($sock);
		} else {
			my $buffer = '';
			if(my $buf = $incoming_bufs{$sock}) {
				my $len = $buf->{'length'};
				my $new_buf = '';
				my $res = sysread($sock, $new_buf, $len - length($buf->{'buf'}));
				if(not $res) {
					&socket_disconnected($sock);
				} else {
					$buf->{'buf'} .= $new_buf;
					if(length($buf->{'buf'}) == $len) {
						$buffer = $buf->{'buf'};
						delete $incoming_bufs{$sock};
					}
				}
			} else {
				my $header = '';
				my $res = sysread($sock, $header, 4);
				if($res != 4) {
					&socket_disconnected($sock);
				} else {
					my $len = unpack('N',$header);
					print STDERR "receiving $len chars...\n";
					my $res = sysread($sock, $buffer, $len);
					if($res != $len) {
						$incoming_bufs{$sock} = {'length' => $len, 'buf' => $buffer};
						$buffer = '';
					}
				}
			}

			&received_packet($sock, $buffer) if $buffer;
		}
	}
}
